### PR TITLE
Minimal support for Ruby 3

### PR DESCRIPTION
**Pull request by PeterJCLaw**
This isn't officially supported by Jekyll and is a bit hacky, so I'm deliberately not documenting the support, however as recent Ubuntu versions now ship with Ruby 3 (not 2.7) and Jekyll isn't showing any signs of moving towards supporting Ruby 3 we're not left with many alternatives.

This works around:
- https://github.com/github/pages-gem/issues/752
- https://github.com/envygeeks/pathutil/pull/5

If we're happy with this fix I'll also apply it to our other Jekyll projects (docs, style, competition-website).

### CODE REVIEW

**Comment by RealOrangeOne** [comment]
As part of this, it might be nice to add Ruby 3 to the CI matrix, so we make sure we keep support. I'm not sure what timeframe you imagine us switching the default install to 3 ( I too share the plague of trying to use older Ruby versions).

**Comment by PeterJCLaw** [comment]
> As part of this, it might be nice to add Ruby 3 to the CI matrix, so we make sure we keep support.

Well that kinda implies it's supported ... which it isn't.

> I'm not sure what timeframe you imagine us switching the default install to 3 ( I too share the plague of trying to use older Ruby versions).

At the moment, none. I don't think we can default to that until the production build (i.e: GitHub Pages itself) switches.

In theory we could move to the manual github pages (via Actions) mechanism, but even then I'm not keen on the production build involving the level of hacks needed here.

**Comment by PeterJCLaw** [comment]
Essentially I see this at the moment as being for those who know what they're doing and want to risk Ruby 3, rather than something we want to encourage people to use.

File: Rakefile [hunk 1]
```diff
@@ -11,6 +11,10 @@ task :dependencies do
     sh('bundle config set --local path "gems"')
   end
   sh('bundle install')
+
+  # Fix pathutil on Ruby 3; works around https://github.com/envygeeks/pathutil/pull/5
+  # as suggested by https://stackoverflow.com/a/73909894/67873
+  sh('sed -i.bak "s/, kwd/, **kwd/" $(bundle exec gem which pathutil)')
```
Comment (sedders123) [inline Rakefile:17]: I don't think `sed` comes as standard on Windows. Can we do this in a cross platform way or only run this fix where `sed` is available?

Comment (PeterJCLaw) [reply Rakefile]: f0e143c _hopefully_ does this.

**Comment by PeterJCLaw** [comment]
Ignore the CircleCI failure -- we don't use CircleCI for this project any more.

**Review by raccube — CHANGES_REQUESTED** [review]

**Comment by raccube** [comment]
We should switch over to building the site ourselves in GitHub Actions, publishing using the [Publish to GitHub Pages action](https://github.com/marketplace/actions/publish-to-github-pages). This would allow us to use a more up to date version of Jekyll.

File: Rakefile [hunk 2]
```diff
@@ -11,6 +13,16 @@ task :dependencies do
...
   sh('bundle install')
+
+  # Fix pathutil on Ruby 3; works around https://github.com/envygeeks/pathutil/pull/5
+  # as suggested by https://stackoverflow.com/a/73909894/67873
+  pathutil_path = `bundle exec gem which pathutil`.chomp()
+  content = File.read(pathutil_path).gsub(', kwd', ', **kwd')
+  backup_path = '#{pathutil_path}.bak'
+  if File.exist?(backup_path) then
+    FileUtil.mv(pathutil_path, backup_path)
+  end
```
[the remark this answers is not in the document]
Comment (PeterJCLaw) [reply Rakefile:24]: Oops, there's a missing negative here -- should only create a backup if one doesn't already exist. However I'm kinda tempted to just drop this anyway. Getting back to the "working" state should be as simple as installing the gem afresh; I don't really expect that most people are going to be reverting to the backup that this creates.

Thoughts?

[the remark this answers is not in the document]
Comment (raccube) [reply Rakefile]: Makes sense to me.

**Review by raccube — APPROVED** [review]

File: Rakefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (PeterJCLaw) [reply Rakefile]: 58d109b08fbb09fa7fca3cc9f784ac8dd1d201a6 drops the backup.

**Pull request merged**